lib/syslog/writer.go

```diff
@@ -244,6 +300,50 @@ func (c bufferedConn) Close() error                { return c.conn.Close() }
 func (c bufferedConn) Flush() error                { return c.buf.Flush() }
 func (c bufferedConn) Write(b []byte) (int, error) { return c.buf.Write(b) }
 
+func cachedDialWriter(network string, address string, config *tls.Config, socksProxy string) (w io.Writer, err error) {
+	if w, ok := writerCache.get(network, address, socksProxy); ok {
+		return w, nil
+	}
+	w, err = syncedDialWriter(network, address, config, socksProxy)
+	if err != nil {
+		writerCache.set(network, address, socksProxy, w)
+	}
+	return w, err
+}
+
+type dialerMutexMap struct {
+	sync.Mutex
+	mu map[string]*sync.Mutex
+}
+
+func (mm dialerMutexMap) lock(network string, address string, socksProxy string) {
```

Comment:
`mm` needs to be a pointer type, otherwise a new copy of the object is done for each method call, which means each goroutine is synchronizing on a local mutex, which is pointless (note that `go vet` would tell you about this, and running a test with the race detector enabled would have caught the issue as well).